File: examples/classes/lazy_attrs.ts
// ClassInstance lazy attributes: fetched from the host on demand.
//
// `lazyAttrs` names cross only when sandbox code reads them — each access
// suspends the sandbox and asks the host. Anything outside the policy raises
// the usual `AttributeError` inside the sandbox.

import assert from 'node:assert/strict'
import { ClassInstance, Monty, MontyRuntimeError } from '@pydantic/monty'

class Config {
  retries = 3
  apiKey = 'hunter2' // never exposed below
}

await using pool = await Monty.create()
await using session = await pool.checkout()
const wrapper = new ClassInstance(new Config(), { lazyAttrs: ['retries'] })
assert.equal(await session.feedRun('cfg.retries', { inputs: { cfg: wrapper } }), 3)

await assert.rejects(session.feedRun('cfg.apiKey', { inputs: { cfg: wrapper } }), (error: unknown) => {
  assert.ok(error instanceof MontyRuntimeError)
  console.log('denied as expected:', error.message)
  return true
})

File: examples/classes/lazy_attrs.py
"""ClassInstance lazy attributes: fetched from the host on demand.

`lazy_attrs` names cross only when sandbox code reads them — each access
suspends the sandbox and asks the host. Anything outside the policy raises
the usual `AttributeError` inside the sandbox.
"""

from pydantic_monty import ClassInstance, Monty, MontyRuntimeError


class Config:
    def __init__(self) -> None:
        self.retries = 3
        self.api_key = 'hunter2'  # never exposed below


with Monty() as pool:
    with pool.checkout() as session:
        wrapper = ClassInstance(Config(), lazy_attrs={'retries'})
        assert session.feed_run('cfg.retries', inputs={'cfg': wrapper}) == 3

    with pool.checkout() as session:
        try:
            session.feed_run('cfg.api_key', inputs={'cfg': ClassInstance(Config(), lazy_attrs={'retries'})})
        except MontyRuntimeError as exc:
            print(f'denied as expected: {exc}')
        else:
            raise AssertionError('expected api_key to be denied')
